### PR TITLE
Optimization: reset bidirectional flag for rules with same src and dst

### PR DESCRIPTION
**Pull request by gozzy**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [ x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [ x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Describe changes:
If one created bidirectional rule like 'alert tcp any any <> any any ...', the rule was checked twice (for each packet in every direction(. This is suboptimal and may give duplicated alerts. To avoid this, bidirectional rules are now checked for the same src and dst (addresses and ports) and if it's the case, the rule is treated as unidirectional and a corresponding message is logged.

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

**Comment by gozzy** [comment]
I don't think that initial behavior was a bug, therefore I've not created a ticked. However I suppose that it slows down packet processing and could be optimized a little bit.

**Comment by gozzy** [comment]
Any thoughts on this?

File: src/detect-parse.c [hunk 1]
```diff
@@ -1796,6 +1796,51 @@ static Signature *SigInitHelper(DetectEngineCtx *de_ctx, const char *sigstr,
...
+    }
+
+    if (!(s->flags & SIG_FLAG_SRC_ANY) || !(s->flags & SIG_FLAG_DST_ANY)) {
+        DetectAddress *src_v4 = s->init_data->src->ipv4_head;
+        DetectAddress *dst_v4 = s->init_data->dst->ipv4_head;
+        DetectAddress *src_v6 = s->init_data->src->ipv6_head;
+        DetectAddress *dst_v6 = s->init_data->dst->ipv6_head;
+
+        if (src_v4 != NULL && dst_v4 != NULL) {
+            if (DetectAddressCmp(src_v4, dst_v4) != ADDRESS_EQ) {
```
Comment (victorjulien) [inline src/detect-parse.c]: both src and dst can be a list of DetectAddress objects, but DetectAddressCmp doesn't compare the list but just a single object.

File: src/detect-parse.c [hunk 2]
```diff
@@ -1796,6 +1796,51 @@ static Signature *SigInitHelper(DetectEngineCtx *de_ctx, const char *sigstr,
     return NULL;
 }
 
+/**
+ * \brief Checks if a signature has the same source and destination
+ * \param s parsed signature
+ *
+ *  \retval 1 if source and destination are the sane, 0 otherwise
+ */
+static int SigHasSameSourceAndDestination(const Signature *s)
```
Comment (victorjulien) [inline src/detect-parse.c]: 'static bool'

File: src/detect-parse.c [hunk 3]
```diff
@@ -1796,6 +1796,51 @@ static Signature *SigInitHelper(DetectEngineCtx *de_ctx, const char *sigstr,
...
+{
+    int ret = 0;
+
+    if (!(s->flags & SIG_FLAG_SP_ANY) || !(s->flags & SIG_FLAG_DP_ANY)) {
+        if (DetectPortCmp(s->sp, s->dp) != 1) {
```
Comment (victorjulien) [inline src/detect-parse.c]: same list issue here

File: src/detect-parse.c [hunk 4]
```diff
@@ -1796,6 +1796,51 @@ static Signature *SigInitHelper(DetectEngineCtx *de_ctx, const char *sigstr,
...
+                goto end;
+            }
+        } else if (!(src_v6 == NULL && dst_v6 == NULL)) {
+            goto end;
+        }
+    }
+
+    ret = 1;
+
+end:
```
Comment (victorjulien) [inline src/detect-parse.c]: no need to use a label and goto's, just return false from each check

**Comment by gozzy** [comment]
Replaced with #3231

**Pull request closed without merging**